connectors/sources/redis.py

```diff
@@ -0,0 +1,283 @@
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import json
+from contextlib import asynccontextmanager
+from enum import Enum
+
+import redis.asyncio as redis
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import hash_id, iso_utc
+
+PAGE_SIZE = 1000
+MAX_POOL_SIZE = 10
+
+
+class KeyType(Enum):
+    HASH = "hash"
+    STREAM = "stream"
+    LIST = "list"
+    SET = "set"
+    ZSET = "zset"
+    JSON = "ReJSON-RL"
+
+
+class RedisClient:
+    """Redis client to handle method calls made to Redis"""
+
+    def __init__(self, configuration):
+        self.configuration = configuration
+        self._logger = logger
+        self.host = self.configuration["host"]
+        self.port = self.configuration["port"]
+        self.database = self.configuration["database"]
+        self.username = self.configuration["username"]
+        self.password = self.configuration["password"]
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    @asynccontextmanager
+    async def client(self):
+        pool = redis.ConnectionPool.from_url(
+            f"redis://{self.username}:{self.password}@{self.host}:{self.port}",
+            decode_responses=True,
+            max_connections=MAX_POOL_SIZE,
+        )
+        yield redis.Redis(connection_pool=pool)
+
+    async def validate_database(self, db, redis_client):
+        try:
+            await redis_client.execute_command("SELECT", db)
+            return await redis_client.ping()
+        except Exception as exception:
+            self._logger.warning(f"Database {db} not found. Error: {exception}")
+
+    async def get_databases(self):
+        """Returns number of databases from config_get response
+
+        Returns:
+            list: List of databases
+        """
+        if self.database != ["*"]:
+            return set(self.database)
+        databases = []
+        try:
+            async with self.client() as redis_client:
+                database = await redis_client.config_get("databases")
+                databases = list(range(int(database.get("databases", 1))))
+        except Exception as exception:
+            self._logger.warning(
+                f"Something went wrong while fetching database list. Error: {exception}",
+                exc_info=True,
+            )
+        return databases
+
+    async def _get_paginated_keys(self, redis_client, pattern, _type=None):
+        """Make a paginated call for fetching database keys.
+
+        Args:
+            redis_client: redis client object
+            pattern: keyname or pattern
+            _type: Datatype for filter data
+
+        Yields:
+            list: list of keys in a database
+        """
+        cursor, keys = await redis_client.scan(
+            match=pattern, count=PAGE_SIZE, _type=_type
+        )
+        yield keys
+
+        while cursor != 0:
```

Comment:
You can consider using [scan_iter](https://redis.readthedocs.io/en/stable/commands.html#redis.commands.core.CoreCommands.scan_iter)